sw/device/lib/runtime/otbn.c

```diff
@@ -83,9 +103,16 @@ otbn_result_t otbn_load_app(otbn_t *ctx, const otbn_app_t app) {
     return kOtbnError;
   }
 
-  if (dmem_size > 0) {
-    if (dif_otbn_dmem_write(&ctx->dif, 0, ctx->app.dmem_start, dmem_size) !=
-        kDifOk) {
+  // Zero all of DMEM
+  otbn_result_t err = otbn_zero_data_memory(ctx);
+  if (err != kOtbnOk) {
+    return err;
+  }
```

Comment:
Oh! If you're doing this then you don't actually have to track the BSS section at all, right? I think this is a totally reasonable approach but, in this case, please can we ditch `dmem_bss_*` from the context object.

Reply:
I think that just means you need the BSS offset, right?

Reply:
I need the BSS offset to do the actual conversion, but `otbn_ptr_to_dmem_addr` should also return an "invalid argument" error if the Ibex pointer passed as input isn't in one of the two Ibex memory sections (data/bss) that point to DMEM for the given app. So to check that the Ibex pointer is valid, I need the bss start/end addresses.

Reply:
Understood, but I'm not sure that feature really matters all that much. I'd probably just suggest you check against the size of the DMEM. That said, if the NOALLOC flag I suggested fixes things, then everything will be contiguous again...

Reply:
I was mistaken in my earlier comment -- I do also need `dmem_bss_start` to compute the DMEM address (because the computation for a bss symbol is basically `ibex_ptr - dmem_bss_start + bss_offset`. I could eliminate `dmem_bss_end` and assume that anything past `dmem_bss_start` and not in `data` range is in bss (because we don't necessarily have the guarantee that `data` will come first if the mask ROM linker script changes!). However, I'm not sure that it would be an improvement; it seems a bit harder to understand than having the same logic for both data and bss, and would only save one field from the context object.

Unfortunately the noalloc suggestion doesn't work, assuming that I understood it correctly -- `set-section-flags` doesn't have an option for `noalloc`, but I tried removing `alloc` from `.bss`, adding `noload`, and putting `NOALLOC` above it in `otbn.ld`; none of these work to reduce the size of the binary. My understanding is that it just gets overridden by the mask ROM linker script, which says `.rodata` is always loaded.

Reply:
Ah, good point about `dmem_bss_start`. At this point I agree: not worth trying to remove!

I'm still slightly confused about the noalloc thing: I rather expected it to work! I'm going to try fiddling around locally a bit and see whether I can get things right.